File: sys/native/src/attention.cpp
#include "mirtal-sys/src/lib.rs.h"

#include <optional>
#include <stdexcept>
#include <string>

namespace mirtal {
namespace {
mx::array sdpa_impl(
    const mx::array& queries,
    const mx::array& keys,
    const mx::array& values,
    float scale,
    const std::string& mask_mode,
    std::optional<mx::array> mask,
    std::optional<mx::array> sinks,
    const mx::Stream& stream) {
  return mx::fast::scaled_dot_product_attention(
      queries, keys, values, scale, mask_mode, mask, sinks, stream);
}
}  // namespace

mx::array sdpa_native(
    const mx::array& queries,
    const mx::array& keys,
    const mx::array& values,
    float scale,
    bool causal,
    const mx::Stream& stream) {
  return sdpa_impl(
      queries,
      keys,
      values,
      scale,
      causal ? "causal" : "",
      std::nullopt,
      std::nullopt,
      stream);
}

std::shared_ptr<Array> sdpa(
    const Array& queries,
    const Array& keys,
    const Array& values,
    const NativeAttentionOptions& options,
    std::shared_ptr<Array> mask,
    std::shared_ptr<Array> sinks,
    const Stream& stream) {
  std::string mask_mode;
  std::optional<mx::array> mask_array;
  switch (options.mask_kind) {
    case 0: break;
    case 1: mask_mode = "causal"; break;
    case 2:
      if (mask == nullptr) throw std::invalid_argument("attention mask is null");
      mask_array = mask->value;
      break;
    default: throw std::invalid_argument("invalid attention mask kind");
  }
  auto sink_array = sinks == nullptr
      ? std::optional<mx::array>{}
      : std::make_optional(sinks->value);
  return std::make_shared<Array>(sdpa_impl(
      queries.value,
      keys.value,
      values.value,
      options.scale,
      mask_mode,
      mask_array,
      sink_array,
      stream.value));
}
}  // namespace mirtal

File: src/attention.rs
use cxx::SharedPtr;
use mirtal_sys::ffi;

use crate::{Array, Error, Graph, Result};

#[derive(Debug, Default, Clone, Copy)]
/// Mask policy for scaled dot-product attention.
pub enum AttentionMask<'array> {
    #[default]
    /// Applies no attention mask.
    None,
    /// Applies MLX's causal attention mask.
    Causal,
    /// Uses a device-resident tensor as the attention mask.
    Array(&'array Array),
}

#[derive(Debug, Clone, Copy)]
/// Options for scaled dot-product attention.
pub struct ScaledDotProductAttention<'array> {
    /// Multiplier applied to query-key scores.
    pub scale: f32,
    /// Mask policy applied before softmax.
    pub mask: AttentionMask<'array>,
    /// Optional device-resident attention-sink values.
    pub sinks: Option<&'array Array>,
}

impl Graph<'_> {
    /// Computes scaled dot-product attention on this graph's stream.
    pub fn scaled_dot_product_attention(
        self,
        queries: &Array,
        keys: &Array,
        values: &Array,
        options: ScaledDotProductAttention<'_>,
    ) -> Result<Array> {
        if !options.scale.is_finite() || options.scale <= 0.0 {
            return Err(Error::InvalidAttention("scale must be finite and positive".into()));
        }
        let (mask_kind, mask) = match options.mask {
            AttentionMask::None => (0, SharedPtr::null()),
            AttentionMask::Causal => (1, SharedPtr::null()),
            AttentionMask::Array(mask) => (2, mask.raw_clone()),
        };
        let sinks = options.sinks.map_or_else(SharedPtr::null, Array::raw_clone);
        let native = ffi::NativeAttentionOptions { scale: options.scale, mask_kind };
        Array::from_raw(
            ffi::sdpa(
                queries.native()?,
                keys.native()?,
                values.native()?,
                &native,
                mask,
                sinks,
                self.native()?,
            )?,
            "scaled dot-product attention",
        )
    }
}
